packages/lit-helpers/src/read-only-properties-mixin.js

```diff
@@ -0,0 +1,90 @@
+/** @typedef {import('lit-element').PropertyDeclaration & { readOnly: Boolean }} PropertyDeclarationWithReadOnly */
+
+/**
+ * @template TBase
+ * @typedef {new (...args: any[]) => TBase} Constructor
+ */
+
+/**
+ * Enables the nofity option for properties to fire change notification events
+ *
+ * @template TBase extends import('lit-element').LitElement
+ * @param {Constructor<TBase>} superclass
+ */
+export const ReadOnlyPropertiesMixin = superclass => {
+  /**
+   * @type {Map<string, symbol>}
+   * @private
+   */
+  const _readOnlyPropertyNamesMap = new Map();
+
+  // @ts-ignore
+  return class ReadOnlyPropertiesClass extends superclass {
+    /**
+     * @inheritdoc
+     * @param  {string} name property name
+     * @param  {PropertyDeclarationWithReadOnly} options augmented property declaration with optional `readOnly` boolean.
+     */
+    static createProperty(name, options) {
+      let finalOptions = options;
+
+      if (options.readOnly) {
+        const privateName = Symbol(name);
+
+        _readOnlyPropertyNamesMap.set(name, privateName);
+
+        Object.defineProperty(this.prototype, name, {
+          get() {
+            return this[privateName];
+          },
+
+          set(value) {
+            // allow for class field initialization
+            /* istanbul ignore if */
+            if (this._readOnlyPropertyInitializedMap.get(name)) return;
+            this[privateName] = value;
+            this._readOnlyPropertyInitializedMap.set(name, true);
+          },
+        });
+
+        finalOptions = { ...options, noAccessor: true };
+      }
+
+      super.createProperty(name, finalOptions);
+    }
+
+    /**
+     * @type {Map<string, boolean>}
+     * @private
+     */
+    _readOnlyPropertyInitializedMap = new Map();
+
+    constructor() {
+      super();
+      this._setPropEntry = this._setPropEntry.bind(this);
+    }
+
+    /**
+     * Set read-only properties
+     * @param  {Object<string, unknown>}  properties object mapping of properties
+     * @private
+     */
+    async setReadOnlyProperties(properties) {
+      await Promise.all(Object.entries(properties).map(this._setPropEntry));
+    }
+
+    /**
+     * @param {[string, unknown]} entry
+     * @return {Promise<unknown>}
+     * @private
+     */
+    _setPropEntry([name, newVal]) {
```

Comment:
Perhaps inline this into `setReadOnlyProperties`? Reduces the footprint of the mixin on the class.